assignments/assignment7/faulty-oops.md

```diff
@@ -0,0 +1,47 @@
+# echo "hello_world" > /dev/faulty
+Unable to handle kernel NULL pointer dereference at virtual address 0000000000000000
+Mem abort info:
+  ESR = 0x0000000096000045
+  EC = 0x25: DABT (current EL), IL = 32 bits
+  SET = 0, FnV = 0
+  EA = 0, S1PTW = 0
+  FSC = 0x05: level 1 translation fault
+Data abort info:
+  ISV = 0, ISS = 0x00000045
+  CM = 0, WnR = 1
+user pgtable: 4k pages, 39-bit VAs, pgdp=0000000041b75000
+[0000000000000000] pgd=0000000000000000, p4d=0000000000000000, pud=0000000000000000
+Internal error: Oops: 0000000096000045 [#1] SMP
+Modules linked in: hello(O) faulty(O) scull(O)
+CPU: 0 PID: 197 Comm: sh Tainted: G           O       6.1.44 #1
+Hardware name: linux,dummy-virt (DT)
+pstate: 80000005 (Nzcv daif -PAN -UAO -TCO -DIT -SSBS BTYPE=--)
+pc : faulty_write+0x10/0x20 [faulty]
+lr : vfs_write+0xc8/0x390
+sp : ffffffc008da3d20
+x29: ffffffc008da3d80 x28: ffffff8001b28000 x27: 0000000000000000
+x26: 0000000000000000 x25: 0000000000000000 x24: 0000000000000000
+x23: 000000000000000c x22: 000000000000000c x21: ffffffc008da3dc0
+x20: 000000558c5edbb0 x19: ffffff8001c01f00 x18: 0000000000000000
+x17: 0000000000000000 x16: 0000000000000000 x15: 0000000000000000
+x14: 0000000000000000 x13: 0000000000000000 x12: 0000000000000000
+x11: 0000000000000000 x10: 0000000000000000 x9 : 0000000000000000
+x8 : 0000000000000000 x7 : 0000000000000000 x6 : 0000000000000000
+x5 : 0000000000000001 x4 : ffffffc000787000 x3 : ffffffc008da3dc0
+x2 : 000000000000000c x1 : 0000000000000000 x0 : 0000000000000000
+Call trace:
+ faulty_write+0x10/0x20 [faulty]
+ ksys_write+0x74/0x110
+ __arm64_sys_write+0x1c/0x30
+ invoke_syscall+0x54/0x130
+ el0_svc_common.constprop.0+0x44/0xf0
+ do_el0_svc+0x2c/0xc0
+ el0_svc+0x2c/0x90
+ el0t_64_sync_handler+0xf4/0x120
+ el0t_64_sync+0x18c/0x190
+Code: d2800001 d2800000 d503233f d50323bf (b900003f) 
+---[ end trace 0000000000000000 ]---
+
+Analysis:
+The kernel Oops occurred because the faulty_write() function in the faulty module attempted to access a NULL pointer, causing the system to crash. The error message "Unable to handle kernel NULL pointer dereference at virtual address 0000000000000000" indicates that the function tried to read or write to an invalid memory location. The call trace shows that this issue originated in faulty_write(), which was triggered when writing to /dev/faulty. This likely happened due to an uninitialized pointer being used. To prevent this, I should add a NULL check before dereferencing the pointer, ensure proper memory allocation using kmalloc(), and use printk() for debugging to catch such issues early.
```

Comment:
It's due to a write of address 0x0, you can use objdump to show you the object content from the faulty module.  You can also use the 0x10/0x20 location referenced in the call stack to locate the offending write in the objdump.